src/shared/watchdog.c

```diff
@@ -210,9 +286,31 @@ int watchdog_setup(usec_t timeout) {
         return r;
 }
 
-usec_t watchdog_runtime_wait(void) {
+int watchdog_setup_pretimeout(usec_t timeout) {
+        /* timeout=0 disables the pretimeout whereas timeout=USEC_INFINITY is a nop. */
+        if ((watchdog_fd >= 0 && timeout == watchdog_pretimeout) || timeout == USEC_INFINITY)
+                return 0;
+
+        /* Initialize the watchdog timeout with the caller value. This value is
+         * going to be updated by update_pretimeout() with the running value,
+         * even if it fails to update the timeout. */
+        watchdog_pretimeout = timeout;
 
-        if (!timestamp_is_set(watchdog_timeout))
+        return update_pretimeout();
+}
+
+static usec_t calc_timeout(void) {
+        /* Calculate the effective timeout which accounts for the watchdog
+         * pretimeout if configured and supported. */
+        if (timestamp_is_set(watchdog_pretimeout) && watchdog_timeout >= watchdog_pretimeout)
+                return usec_sub_unsigned(watchdog_timeout, watchdog_pretimeout);
```

Comment:
you already checked for overflow here, hence no need to use usec_sub_unsigned(), since the only thing that really does is do overflow check/saturate, a second time. just do `return wacthdog_timeout - watchdog_pretimeout` here.